Update docs, add docs for bundling and Babbage features (#1245)
Plutonomicon/cardano-transaction-lib: doc/babbage-features.md

# CTL Babbage feature overview

This document is a reference/explainer for the new CTL APIs introduced for Babbage support.

## Reference Inputs

[Reference inputs](https://cips.cardano.org/cips/cip31/#referenceinputs) allow looking at an output without spending it in Plutus scripts.

> In CTL reference inputs can be used by providing a value of type `InputWithScriptRef` that specifies whether an output should be spent or referenced. One of the constraints that use it is `mustSpendScriptOutputUsingScriptRef`.

Reviewer: It's better to mention the `mustReferenceOutput` constraint there, since it allows to simply add a reference input without using it to satisfy script witnessing requirements. Though, we should still leave a note that the reference inputs added using this constraint cannot be used to resolve scripts and also mention `mustSpendScriptOutputUsingScriptRef` and its variants that should be used for this purpose.

> In CTL reference inputs can be used by providing a value of type `InputWithScriptRef` that specifies whether an output should be spent or referenced. One of the constraints that use it is `mustSpendScriptOutputUsingScriptRef`.

Reviewer 2: ```suggestion
There are two ways to use an input as a reference in the constraints API. The first is via `mustReferenceOutput`, which allows Plutus scripts to access the datum and value of the output. The second is by providing constraints which accept a value of the type `InputWithScriptRef` with the `RefInput` constructor. These allow scripts (validating or minting) to be reused by reference between multiple transactions without including them in those transactions, explained further in [Reference Scripts](#reference-scripts).
```

[Usage example](../examples/PlutusV2/ReferenceInputs.purs)

## Reference Scripts

> [Reference Scripts](https://developers.cardano.org/docs/governance/cardano-improvement-proposals/cip-0033/) allow to use the scripts without attaching them to the transaction (and using a reference instead).

Reviewer 2: ```suggestion
[Reference Scripts](https://developers.cardano.org/docs/governance/cardano-improvement-proposals/cip-0033/) allows the use of scripts without attaching them to the transaction (and using a reference instead).
```

> In CTL, reference scripts can be utilized by first creating a reference point for the script to be used later via `mustPayToScriptWithScriptRef` (or its variants).

Reviewer 2: ```suggestion
Reference scripts can be utilized in CTL by first creating a reference point for the script to be used later via `mustPayToScriptWithScriptRef` (or its variants).
```

This constraint utilises a new `ScriptRef` type that includes either a native script or a Plutus script.

> Then, `mustSpendScriptOutputUsingScriptRef` (or its variants) can be used to use a reference script. It accepts a value of type `InputWithScriptRef` that allows to either spend the UTxO, or just reference it.

Reviewer: ```suggestion
Then, `mustSpendScriptOutputUsingScriptRef` (or its variants) can be used to use a reference script. It accepts a value of type `InputWithScriptRef` that specifies whether the UTxO with the reference script should be spent or referenced.
```

[Usage example](../examples/PlutusV2/ReferenceScripts.purs)

## Inline Data

> [CIP-32](https://developers.cardano.org/docs/governance/cardano-improvement-proposals/cip-0032/) introduces inline data feature that allows to store datum values directly in transaction outputs, instead of storing just the hashes.

Reviewer 2: ```suggestion
[CIP-32](https://developers.cardano.org/docs/governance/cardano-improvement-proposals/cip-0032/) introduces the inline data feature that allows storing datum values directly in transaction outputs, instead of just the hashes.
```

In CTL, alternating between datum storage options can be achieved by specifying a `DatumPresence` value with constraints that accept it, like `mustPayToPubKeyWithDatum`.

[Usage example](../examples/PlutusV2/InlineDatum.purs)

## Collateral Output

> To trigger a [collateral return](https://cips.cardano.org/cips/cip40/), `mustNotBeValid` constraint should be explicitly specified (otherwise a script error would be detected earlier and the transaction will not be sent).

Reviewer: Not sure if this is worth mentioning here, but some browser wallets (Nami for sure) will not allow to sign an invalid transaction.

> To trigger a [collateral return](https://cips.cardano.org/cips/cip40/), `mustNotBeValid` constraint should be explicitly specified (otherwise a script error would be detected earlier and the transaction will not be sent).

Reviewer 2: ```suggestion
[CIP-40](https://cips.cardano.org/cips/cip40/) introduces explicit collateral output. On validation failure, previously the entire collateral was consumed. Now, if excess collateral is supplied, even with native assets, the surplus can be returned on validation failure.

Collateral output is automatically added to transactions in CTL. To trigger a collateral return, the `mustNotBeValid` constraint should be explicitly specified, otherwise a script error would be detected earlier and the transaction will not be sent.
```
